crates/fuel-core/src/state/historical_rocksdb.rs

```diff
@@ -0,0 +1,787 @@
+    #[default]
+    /// The checkpoint will be created only for the latest height.
+    NoRewind,
+    /// The checkpoint will be created for each height.
+    RewindFullRange,
+    /// The checkpoint will be created for each height
+    /// in the range `[latest_height-size..latest_height]`.
+    RewindRange { size: NonZeroU64 },
+}
+
+#[derive(Debug)]
+pub struct HistoricalRocksDB<Description> {
+    state_rewind_policy: StateRewindPolicy,
+    db: RocksDb<Description>,
+    history: RocksDb<Historical<Description>>,
+}
+
+impl<Description> HistoricalRocksDB<Description>
+where
+    Description: DatabaseDescription,
+{
+    pub fn new(
+        db: RocksDb<Description>,
+        state_rewind_policy: StateRewindPolicy,
+    ) -> DatabaseResult<Self> {
+        let path = db.path().join("history");
+        let history = RocksDb::default_open(path, None)?;
+        Ok(Self {
+            state_rewind_policy,
+            db,
+            history,
+        })
+    }
+
+    fn reverse_history_changes(&self, changes: &Changes) -> StorageResult<Changes> {
+        let mut reverse_changes = Changes::default();
+
+        for (column, column_changes) in changes {
+            let results = self
+                .db
+                .multi_get(*column, column_changes.iter().map(|(k, _)| k))?;
+
+            let entry = reverse_changes
+                .entry(*column)
+                .or_insert_with(Default::default);
+
+            for (was, (key, became)) in results.into_iter().zip(column_changes.iter()) {
+                match (was, became) {
+                    (None, WriteOperation::Remove) => {
+                        // Do nothing since it was not existing, and it was removed.
+                    }
+                    (None, WriteOperation::Insert(_)) => {
+                        entry.insert(key.clone(), WriteOperation::Remove);
+                    }
+                    (Some(old_value), WriteOperation::Remove) => {
+                        entry.insert(
+                            key.clone(),
+                            WriteOperation::Insert(old_value.into()),
+                        );
+                    }
+                    (Some(old_value), WriteOperation::Insert(new_value)) => {
+                        if old_value.as_slice() != new_value.as_slice() {
+                            entry.insert(
+                                key.clone(),
+                                WriteOperation::Insert(old_value.into()),
+                            );
+                        }
+                    }
+                }
+            }
+        }
+        Ok(reverse_changes)
+    }
+
+    pub fn latest_view(&self) -> StorageResult<RocksDb<Description>> {
+        let db_view = self.db.create_snapshot();
+        Ok(db_view)
+    }
+
+    pub fn create_view_at(
+        &self,
+        height: &Description::Height,
+    ) -> StorageResult<ViewAtHeight<Description>> {
+        let latest_view = self.latest_view()?;
```

Comment:
Should we do some kind of consistency check here? Could there be a race condition where the latest view snapshot is not in sync with the historical database when the snapshots are created?

Reply:
This situation is handled on the `GenericDatabase` level. When we commit to the database, we lock the current block height.

<img width="829" alt="image" src="https://github.com/FuelLabs/fuel-core/assets/18346821/f6f65d00-80fe-4617-9e16-f762ca78dea8">

The same is in the case of `view_at.` We lock the block height so the view can only be created outside of the commit.

<img width="993" alt="image" src="https://github.com/FuelLabs/fuel-core/assets/18346821/0f90222f-282d-48ae-8792-9efc1773b84b">